lib/scholar/metrics.ex

```diff
@@ -332,28 +351,58 @@ defmodule Scholar.Metrics do
       iex> y_pred = Nx.tensor([0, 1, 0, 2, 2, 2], type: {:u, 32})
       iex> Scholar.Metrics.confusion_matrix(y_true, y_pred, num_classes: 3)
       #Nx.Tensor<
-        u64[3][3]
+        f32[3][3]
+        [
+          [1.0, 1.0, 0.0],
+          [1.0, 0.0, 1.0],
+          [0.0, 0.0, 2.0]
+        ]
+      >
+
+      iex> y_true = Nx.tensor([0, 0, 1, 1, 2, 2], type: {:u, 32})
+      iex> y_pred = Nx.tensor([0, 1, 0, 2, 2, 2], type: {:u, 32})
+      iex> sample_weights = [2, 5, 1, 1.5, 2, 8]
+      iex> Scholar.Metrics.confusion_matrix(y_true, y_pred, num_classes: 3, sample_weights: sample_weights, normalize: :predicted)
+      #Nx.Tensor<
+        f32[3][3]
         [
-          [1, 1, 0],
-          [1, 0, 1],
-          [0, 0, 2]
+          [0.6666666865348816, 1.0, 0.0],
+          [0.3333333432674408, 0.0, 0.1304347813129425],
+          [0.0, 0.0, 0.8695651888847351]
         ]
       >
   """
   deftransform confusion_matrix(y_true, y_pred, opts \\ []) do
-    confusion_matrix_n(y_true, y_pred, NimbleOptions.validate!(opts, @general_schema))
+    confusion_matrix_n(y_true, y_pred, NimbleOptions.validate!(opts, @confusion_matrix_schema))
   end
 
   defnp confusion_matrix_n(y_true, y_pred, opts) do
     check_shape(y_pred, y_true)
+    weights = validate_weights(opts[:sample_weights], Nx.axis_size(y_true, 0))
+    weights = if Nx.rank(weights) == 0, do: Nx.broadcast(weights, y_true), else: weights
```

Comment:
Validate weights already does this broadcast, no?

Reply:
No, since validate weights return scalar 1.0 if weights are initially `nil`

Reply:
https://github.com/elixir-nx/scholar/blob/main/lib/scholar/shared.ex#L54-L56

Reply:
Exactly. Instead of having it as `nil`, check the input types in `deftransform` and make it so the default value is either `1` or `1.0`. This way you get the integer/float behaviour and you can also remove the broadcast. :)